clients/gtest/rocblas_gtest_main.cpp

```diff
@@ -209,43 +233,71 @@ static void rocblas_test_sigaction()
     act.sa_flags   = 0;
     act.sa_handler = rocblas_test_signal_handler;
 
-    for(int sig : {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGPIPE, SIGSEGV, SIGSYS, SIGUSR1, SIGUSR2})
-    {
+    for(int sig : {SIGABRT, SIGBUS, SIGFPE, SIGILL, SIGALRM, SIGSEGV, SIGSYS, SIGUSR1, SIGUSR2})
         sigaction(sig, &act, nullptr);
-    }
 }
 
-// Lambda wrapper which detects signals and exceptions in an invokable function
-void catch_signals_and_exceptions_as_failures(const std::function<void()>& test)
+// Catch std::terminate
+static void terminate_handler()
 {
-    // Save this thread's id, to detect signals in different threads
-    rocblas_test_sighandler_tid = pthread_self();
+    if(state.rocblas_sighandler_enabled)
+        siglongjmp(state.rocblas_sigjmp_buf, -1);
+}
 
-    // Set up the return point
-    int sig = sigsetjmp(rocblas_sigjmp_buf, true);
+// Lambda wrapper which detects signals and exceptions in an invokable function
+void catch_signals_and_exceptions_as_failures(const std::function<void()>& test, bool set_alarm)
+{
+    // Save the current state (to allow nested calls to this function)
+    // Variables must be volatile to survive across sigsetjmp
+    volatile auto old_state     = state;
+    volatile auto old_terminate = std::set_terminate(terminate_handler); // Catch std::terminate
 
-    // If this is a return, indicate the signal received
-    if(sig)
-    {
-        rocblas_sighandler_enabled = false;
-        FAIL() << "Received " << strsignal(sig) << " signal";
-    }
-    else
+    // Set up the return point, and handle siglongjmp returns back to here
+    switch(sigsetjmp(state.rocblas_sigjmp_buf, true))
     {
+    case 0: // Normal fall-through of sigsetjmp
+
+        // Alarm to detect deadlocks or hangs
+        if(set_alarm)
+            alarm(TEST_TIMEOUT);
+
+        // Save this thread's id, to detect signals in different threads
+        state.rocblas_test_sighandler_tid = pthread_self();
+
+        // Enable the signal handler
+        state.rocblas_sighandler_enabled = true;
+
         // Run the test function, catching signals and exceptions
-        // Disable the signal handler after running the test
-        rocblas_sighandler_enabled = true;
         try
         {
             test();
-            rocblas_sighandler_enabled = false;
         }
         catch(...)
         {
-            rocblas_sighandler_enabled = false;
-            FAIL() << "Received unhandled exception";
+            FAIL() << "Received uncaught exception";
         }
+        break;
+
+    case 1: // signal_received indicates the signal which was received
+        FAIL() << "Received " << strsignal(signal_received) << " signal";
+        break;
+
+    case -1: // terminate_handler()
+        FAIL() << "Received std::terminate";
+        break;
+
+    default: // Unknown return value
+        FAIL() << "Internal error in rocblas_gtest_main.cpp";
+        break;
```

Comment:
`sigsetjmp()` can return `0` after a normal setup call, `1` if `siglongjmp` was used with an argument of `1` to return, `-1` if `terminate_handler` was used. Other than those values, it's an internal error.